### PR TITLE
Qualification tool: Operator mapping from plugin to CSV file

### PR DESCRIPTION
**Pull request by nartal1**
This fixes https://github.com/NVIDIA/spark-rapids/issues/5263

In this PR we are creating csv files for
1. Execs supported.
2. Expressions supported
3. Operator score - Currently we are assinging static scores(speed up) for execs and expressions. We need to figure out a better way to publish the speed for each execs/expressions.
This data can be used by qualification tool for qualifing the apps. Added unit tests.

### CODE REVIEW

**Comment by nartal1** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 1]
```diff
@@ -2147,11 +2148,28 @@ object SupportedOpsDocs {
...
 
-object SupportedOpsForTools {
+object SupportedOpsForTools extends Logging {
 
   private lazy val allSupportedTypes =
     TypeSigUtil.getAllSupportedTypes()
 
+  // if a string contains what we are going to use for a delimiter, replace
+  // it with something else
+  def replaceDelimiter(str: String, delimiter: String): String = {
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:2158]: make private

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 2]
```diff
@@ -2193,15 +2211,119 @@ object SupportedOpsForTools {
...
-  def help(): Unit = {
-    outputSupportIO()
+  private def operatorMappingWithScore(): Unit = {
+    val header = Seq("CPUOperator", "Score")
+    println(header.mkString(","))
+    GpuOverrides.execs.values.toSeq.sortBy(_.tag.toString).foreach { rule =>
+      val checks = rule.getChecks
+      if (rule.isVisible && checks.forall(_.shown)) {
+        val cpuName = rule.tag.runtimeClass.getSimpleName
+        val allCols = Seq(cpuName, "2")
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:2221]: add comment here why we set to 2.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 3]
```diff
@@ -2193,15 +2211,119 @@ object SupportedOpsForTools {
...
+        println(s"${allCols.map(replaceDelimiter(_, ",")).mkString(",")}")
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:2222]: I don't think we need replaceDelimiter here, the class name should be a single word, same with below

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 4]
```diff
@@ -2193,15 +2211,119 @@ object SupportedOpsForTools {
...
+      if (rule.isVisible && checks.forall(_.shown)) {
+        val cpuName = rule.tag.runtimeClass.getSimpleName
+        val allCols = Seq(cpuName, "3")
+        println(s"${allCols.map(replaceDelimiter(_, ",")).mkString(",")}")
+      }
+    }
+  }
+
+  private def outputSupportedExecs(): Unit = {
+    // TODO Look at what we have for defaults for some configs because if the configs are off
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:2237]: we should check the rule.disabledMsg.isDefined to see if they are disabled by default (CollectLimitExec is one example).  

We have other configs to control certain execs like ENABLE_LEFT_OUTER_JOIN, lets file a follow up to figure out a good way to account for those.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 5]
```diff
@@ -2193,15 +2211,119 @@ object SupportedOpsForTools {
...
+        val firstCol = Seq(rule.tag.runtimeClass.getSimpleName)
+        val staticCols = Seq(sqlFunctions.getOrElse(" "), rule.notes().getOrElse("None"))
+
+        representative.foreach {
+          case (context, data) =>
+            data.keys.foreach { param =>
+              val supportLevelOps = allSupportedTypes.toSeq.map { t =>
+                allData(t)(context)(param).text
+              }
+              val isSupportedExec = Seq(if (supportLevelOps.forall(_.equals("NS"))) "NS" else "S")
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:2298]: nit: isSupportedExpr

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 6]
```diff
@@ -2193,15 +2211,119 @@ object SupportedOpsForTools {
...
+      }
+    }
+  }
+
+  def help(printType: String): Unit = {
+    printType match {
+      case a if a.equals("execs") => outputSupportedExecs()
+      case expr if (expr.equals("exprs")) => outputSupportedExpressions()
+      case score if (score.equals("operatorScore")) => operatorMappingWithScore()
+      case _ => outputSupportIO()
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:2313]: change to match on "ioOnly" and then add a default case which throws an error and prints out valid options.

File: tools/src/main/resources/supportedExecs.csv [hunk 7]
```diff
@@ -0,0 +1,47 @@
+Exec,Supported,Notes,Params,BOOLEAN,BYTE,SHORT,INT,LONG,FLOAT,DOUBLE,DATE,TIMESTAMP,STRING,DECIMAL,NULL,BINARY,CALENDAR,ARRAY,MAP,STRUCT,UDT
+CoalesceExec,S,None,Input/Output,S,S,S,S,S,S,S,S,PS,S,S,S,NS,NS,PS,PS,PS,NS
+CollectLimitExec,S,This is disabled by default because Collect Limit replacement can be slower on the GPU; if huge number of rows in a batch it could help by limiting the number of rows transferred from GPU to CPU,Input/Output,S,S,S,S,S,S,S,S,PS,S,S,S,NS,NS,PS,PS,PS,NS
```
Comment (tgravescs) [inline tools/src/main/resources/supportedExecs.csv:3]: this is off by default so once we add that logic shouldn't be S

File: tools/src/main/resources/supportedExprs.csv [hunk 8]
```diff
@@ -0,0 +1,643 @@
+Expression,Supported,SQL Func,Notes,Context,Params,BOOLEAN,BYTE,SHORT,INT,LONG,FLOAT,DOUBLE,DATE,TIMESTAMP,STRING,DECIMAL,NULL,BINARY,CALENDAR,ARRAY,MAP,STRUCT,UDT
+Abs,S,`abs`,None,project,input,NA,S,S,S,S,S,S,NA,NA,NA,S,NA,NA,NA,NA,NA,NA,NA
```
Comment (tgravescs) [inline tools/src/main/resources/supportedExprs.csv:2]: did you load this up in Sark to make it was valid csv and no extra delimiters?  If not please check it

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala [hunk 9]
```diff
@@ -52,12 +61,131 @@ class PluginTypeChecker {
...
+            " must have at least the header and one line")
+      }
+      // first line is header
+      val header = fileContents.head.split(",").map(_.toLowerCase)
+      // the rest of the rows are file with type supported info
+      fileContents.tail.foreach { line =>
+        val cols = line.split(",")
+        if (header.size != cols.size) {
+          throw new IllegalStateException("operatorsScore file appears corrupt," +
+              " header length doesn't match rows length")
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala:121]: print the row that doesn't match

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala [hunk 10]
```diff
@@ -52,12 +61,131 @@ class PluginTypeChecker {
...
+        }
+        val operator = cols(0)
+        val score = cols(1).toInt
+        supportedOperatorsScore.put(operator, score)
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala:125]: the 2 above variables seem unneeded just put:
supportedOperatorsScore.put(cols(0), cols(1).toInt)

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala [hunk 11]
```diff
@@ -52,12 +61,131 @@ class PluginTypeChecker {
...
+            " must have at least the header and one line")
+      }
+      // first line is header
+      val header = fileContents.head.split(",").map(_.toLowerCase)
+      // the rest of the rows are execs with type supported info
+      fileContents.tail.foreach { line =>
+        val cols = line.split(",")
+        if (header.size != cols.size) {
+          throw new IllegalStateException("supportedExecs file appears corrupt," +
+              " header length doesn't match rows length")
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala:148]: print row

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala [hunk 12]
```diff
@@ -52,12 +61,131 @@ class PluginTypeChecker {
...
+        }
+
+        val operator = cols(0)
+        val isSupported = cols(1)
+        supportedExecs.put(operator, isSupported)
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala:153]: same comment as above, don't need the extra vals

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala [hunk 13]
```diff
@@ -52,12 +61,131 @@ class PluginTypeChecker {
...
+            " must have at least the header and one line")
+      }
+      // first line is header
+      val header = fileContents.head.split(",").map(_.toLowerCase)
+      // the rest of the rows are exprs with type supported info
+      fileContents.tail.foreach { line =>
+        val cols = line.split(",")
+        if (header.size != cols.size) {
+          throw new IllegalStateException("supportedExprs file appears corrupt," +
+              " header length doesn't match rows length")
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala:176]: print row

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala [hunk 14]
```diff
@@ -52,12 +61,131 @@ class PluginTypeChecker {
...
+        supportedOperatorsScore.put(operator, score)
+      }
+    } finally {
+      source.close()
+    }
+    supportedOperatorsScore.toMap
+  }
+
+  private def readSupportedExecs(source:BufferedSource): (Map[String, String]) = {
+    val supportedExecs = HashMap.empty[String, String]
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala:134]: these 3 functions are basically all the same code, perhaps we can combine them. The operators score one we could just do a final map over it to convert the score to Int after the fact.

File: tools/src/test/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeCheckerSuite.scala [hunk 15]
```diff
@@ -97,4 +97,34 @@ class PluginTypeCheckerSuite extends FunSuite with Logging {
...
+      val supText = (header + "FilterExec,3\n").getBytes(StandardCharsets.UTF_8)
+      val csvSupportedFile = Paths.get(outpath.getAbsolutePath, "testScore.txt")
+      Files.write(csvSupportedFile, supText)
+      checker.setOperatorScore(csvSupportedFile.toString)
+      val operScore = checker.getOperatorScore
+      assert(operScore.contains("FilterExec"))
+      assert(!operScore.contains("ProjectExec"))
+      assert(operScore("FilterExec") == 3)
+    }
+  }
```
Comment (tgravescs) [inline tools/src/test/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeCheckerSuite.scala:114]: add one for operator score that reads the default file as well.

File: tools/src/test/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeCheckerSuite.scala [hunk 16]
```diff
@@ -97,4 +97,34 @@ class PluginTypeCheckerSuite extends FunSuite with Logging {
...
+
+  test("supported Execs") {
+    val checker = new PluginTypeChecker
+    val result = checker.getSupportedExecs
+    assert(result.contains("ShuffledHashJoinExec"))
+    assert(result("ShuffledHashJoinExec") == "S")
```
Comment (tgravescs) [inline tools/src/test/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeCheckerSuite.scala:120]: if we change it so CollectLimitExec is NS, check that here

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 17]
```diff
@@ -2147,11 +2148,28 @@ object SupportedOpsDocs {
   }
 }
 
-object SupportedOpsForTools {
+object SupportedOpsForTools extends Logging {
```
Comment (tgravescs) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:2151]: I dont' think this is used anymore, remove the extends Logging

File: tools/src/main/resources/supportedExprs.csv (hunk 8 above)
Comment (nartal1) [reply tools/src/main/resources/supportedExprs.csv:2]: Yes, I was able to read this and other csv files to a dataframe correctly.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala (hunk 4 above)
Comment (nartal1) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:2240]: Filed follow on issue - https://github.com/NVIDIA/spark-rapids/issues/5305

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala [hunk 18]
```diff
@@ -52,12 +61,74 @@ class PluginTypeChecker {
...
+  }
+
+  def setSupportedExecs(filePath: String): Unit = {
+    val source = Source.fromFile(filePath)
+    supportedExecs = readSupportedOperators(source)
+  }
+
+  def setSupportedExprs(filePath: String): Unit = {
+    val source = Source.fromFile(filePath)
+    supportedExprs = readSupportedOperators(source)
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala:76]: make comments here and in the Execs functions that we are only reading the first 2 columns for now, other columns ignored intentionally

**Review by tgravescs — APPROVED** [review]

**Pull request merged**